Refactor initial conditions

TimothyWillard opened the pull request:
### Describe your changes.

This pull request... 

- [x] Moves the `initial_conditions` submodule to a subpackage.
- [x] Adds `ModelMeta` class to represent the core metadata of a configuration file and do basic filesystem operations.
- [x] Uses `ModelMeta` in `ModelInfo` to validate meta & setup filesystem.
- [x] Provide `ModelMeta` to `InitialConditions.from_confuse_config` to enable filesystem access.
- [x] Implement "SetInitialConditions", "SetInitialConditionsFolderDraw", "FromFile", "InitialConditionsFolderDraw" methods as `FileOrFolderDrawInitialConditions`.
- [x] Unit test/document `DefaultInitialConditions`.
- [x] ~Unit test/document `FileOrFolderDrawInitialConditions`.~ Follow up  in #593.
- [x] Integrate new initial conditions into existing `gempyor` codebase.
- [x] Explore options for incorporating parameters into the initial conditions methods ~(might be better for a follow up PR)~.
- [x] Refactor `_read_initial_condition_from_tidydataframe`.
- [x] Unit test `_read_initial_condition_from_tidydataframe`.
- [x] Refactor `_read_initial_condition_from_seir_output`. (~Kinda, compartment filtering functionality needs to be extracted~).
- [x] Unit test `_read_initial_condition_from_seir_output`.
- [x] ~Integrate with flu configuration writer script. @MacdonaldJoshuaCaleb to point to relevant script(s) here.~ Moved to #597.
- [x] ~Add initial conditions method for specifying directly in a configuration file.~ Moved to #594.
- [x] Simplify `path_prefix` handling, should be obtainable from the `meta.path_prefix` attribute.
- [x] Add end user documentation to GitBook.
- [x] Pull a89fd86be78970b98f6b5d535892b3884027fccc (`nlqmtrru`) into the correct commit, ultimately it's a mistake.
- [x] Add functionality to parse the arguments of `create_initial_conditions` so parameters are provided in a more user friendly way.
- [x] ~Extract the initial conditions module loading from `initial_conditions_from_plugin` into an internal helper (thinking ahead to extracting/generalizing the plugin system for the rest of gempyor).~ Based handled by follow up work in #421.
- [x] Add examples to `_inspect_requested_parameters` docstring.
- [x] Address multiprocessing pickling error that occurs when using initial conditions plugin with EMCEE inference.
- [x] Clean up 0baf9cdffec2fd3b9bf1eb226bba7ef40b7dfd0d (oxqlonpk) containing the fix for ^.
- [x] Test initial conditions plugins with R inference.
- [x] Add to documentation a note about what occurs when user requests a parameter that does not exist, i.e. a typo (https://github.com/HopkinsIDD/flepiMoP/pull/572#discussion_r2223758429).
- [x] ~Rethink current pydantic usage, users would prefer to receive configuration as arguments rather than as attributes to a class (https://github.com/HopkinsIDD/flepiMoP/pull/572#discussion_r2223764197).~ Outside the scope of this PR, likely best for a general purpose plugin system.
- [x] Confirm if initial conditions and seeding are mutually exclusive, I do not think so (https://github.com/HopkinsIDD/flepiMoP/pull/572#discussion_r2223707292).
- [x] ~Allow seeding to also be parameterized similarly to initial conditions in this PR (https://github.com/HopkinsIDD/flepiMoP/pull/572#discussion_r2223707292).~ Likely best for follow up work in #587.
- [x] Confirm the behavior of seeding, re: https://github.com/HopkinsIDD/flepiMoP/pull/572#discussion_r2223715822.
- [x] ~Investigate functional definitions for initial condition plugins rather than the current subclassing approach (https://github.com/HopkinsIDD/flepiMoP/pull/572#discussion_r2223724792).~ Best handled by follow up work in #421.
- [x] Support more than time series or distribution parameters being requested by plugins (https://github.com/HopkinsIDD/flepiMoP/pull/572#discussion_r2223724792).

### Does this pull request make any user interface changes? If so please describe.

The user interface changes are none for builtin methods, but minor changes to how plugins work with additional features.

Those are reflected in updates to the documentation in `documentation/gitbook/gempyor/model-implementation/specifying-initial-conditions.md`.

TimothyWillard committed:
First pass refactoring initial conditions

Made a first pass refactoring initial conditions by address most, but
not all, of the pylint suggestions. Includes reordering of logic,
addition of docstrings, code formatting, etc. but importantly no changes
to logic. Other larger changes include:

* Renamed the `InitialConditionsFactory` function to 
  `initial_conditions_factory` to fit pythonic coding style.
* Unit tests and minor refactoring of the `check_population` internal
  validation utility, namely simplifying input arguments.
* Added unit tests for the `read_initial_condition_from_tidydataframe`
  helper function for parsing ICs from a "tidy" data frame.

TimothyWillard committed:
Move ICs from submodule to subpackage

Moved initial conditions from being a submodule of `gempyor` to being a
`subpackage`. Results in no net interface changes, opens the door for
more invasive refactoring.

TimothyWillard committed:
Split initial conditions functionality out

TimothyWillard committed:
Simplify `check_population` internals

Simplifed the internals of `check_population` by leaning on numpy's
functionality more.

TimothyWillard committed:
Draft modular initial conditions

First pass at modular initial conditions, enabled by the
`InitialConditionsABC` that defines the interface for an initial
conditions module. Also implmented `DefaultInitialConditions` as an
example of how to use said ABC.

TimothyWillard committed:
Add `ModelMeta` class for core model metadata

Added `gempyor.model_meta.ModelMeta` class as a data container for core
model metadata and providing basic filesystem operations independent of
the `ModelInfo` mega class.

Use `ModelMeta` class to validate model metadata for `ModelInfo` and do
some light filesystem setup. Currently reassigning attributes from
`ModelMeta` back to `ModelInfo` for backwards compatibility.

TimothyWillard committed:
Extract `TimeSetup` from `gempyor.model_info`

Extracted the `TimeSetup` class from `gempyor.model_info` and placed it
into the new `gempyor.time_setup` module and also refactored to use a
pydantic representation instead of the previous plain class
representation.

TimothyWillard committed:
Initial `FileOrFolderDrawInitialConditions` draft

First draft of
`gempyor.initial_conditions.FileOrFolderDrawInitialConditions`. Limited
on documentation and unit tests, but provides the
'SetInitialConditions', 'SetInitialConditionsFolderDraw', 'FromFile',
and 'InitialConditionsFolderDraw' initial condition methods. Also had to
take modified versions of helper functions from `gempyor._readers` that
need to be heavily refactored.

jc-macdonald commented:
some thoughts - we do need to be able to treat initial conditions (or at least S(0) and R(0)) as special in the sense that they're not independent, so we should think carefully about how to handle this.  I(0) and H(0) are also clearly interlinked, really only V(0) can probably be argued to be fully independent - we can probably get around that to an extent with some binomials, but dirchlet is the natural distribution if we want to do it "right".

TimothyWillard committed:
Replace `InitialConditions` with modular plugins

Replaced the previous `InitialConditions` monolithic class with modular
classes that subclass `InitialConditionsABC`. These subclasses then
register themselves as plugins via `register_initial_conditions_plugin`.
These are then accessible via the `initial_conditions_from_plugin`
function. Initial condition plugins can then be reset using the
`_reset_initial_conditions_plugins` helper, mostly for unit testing
purposes.

Furthermore, this has been integrated directly into the rest of
`gempyor`. The `get_initial_conditions_data` method was added to the
model info object to more easily access initial conditions, similar to
how seeding is done.

TimothyWillard committed:
Remove unused `gempyor.initial_conditions._readers`

Removed an unused helper module that contained utilities for the prior
initial conditions implementation. Those utilities have been moved into
`gempyor.initial_conditions._file_or_folder_draw` as internal helpers.
Also made corresponding moves/edits to the unit tests.

TimothyWillard committed:
Add unit tests for `DefaultInitialConditions`

In particular testing the outputs of the `create_initial_conditions`
method provided by
`gempyor.initial_conditions.DefaultInitialConditions`.

TimothyWillard committed:
Add unit tests for `initial_conditions_from_plugin`

Added unit tests for the function used to generate an
`InitialConditionsABC` instance, namely:
* Checking that `DefaultInitialConditions` is returned when 'method' is
  not specified, with a warning, and
* A `ValueError` is raised when the 'specified' method is not found in
  the registered initial condition plugins.

TimothyWillard committed:
Make use of `path_prefix` from `ModelMeta`

Rather than requiring an explicit path prefix value be provided make use
of the 'path_prefix' attribute from `gempyor.model_meta.ModelMeta` for
filesystem interactions.

TimothyWillard committed:
Start `FileOrFolderDrawInitialConditions` tests

Initial unit tests for the
`gempyor.initial_conditions.FileOrFolderDrawInitialConditions` class.
Began primarily with initialization errors and warnings.

TimothyWillard committed:
Edit `_read_initial_condition_from_tidydataframe`

* Added `subset_dataframe` to `gempyor.compartments.Compartments` for
  looping over a dataframe subsetting by compartments,
* Added internal utility `_invert_into_dict` for efficiently inserting
  one value to many keys in a dict,
* Allow 'rest' allocations to work when using the 'Set*' methods for
  initial conditions, and
* Refactored `read_initial_condition_from_seir_output` to also use
  `Compartments.subset_dataframe`.

TimothyWillard committed:
Provide parameters to initial conditions

Made modifications to the
`InitialConditionsABC.create_initial_conditions` abstract method to
allow it to accept an instance of `Paramters` (represents information
about SEIR parameters in the abstract) and a numpy array of actualized
parameter values. Subsequent changes were made to:

* The two implementations of `InitialConditionsABC`,
* The `InitialConditionsABC.get_initial_conditions` wrapper method,
* The `ModelInfo.get_initial_conditions_data` wrapper method, and
* The various calls to `get_initial_conditions_data` in the seir unit
  tests and `gempyor.seir/inference` modules.

Other minor test adjustments were made to `tests/initial_conditions/` to
provide dummy values for those parameters. Unused by current
implementations.

TimothyWillard committed:
Add `gempyor.utils.ConfigurationWarning`

Added a custom warning for potential configuration file issues. This
warning is desirned to indicate configuration issues that do not prevent
flepiMoP from running but may produce unexpected behavior or that the
user made a mistake of some kind.

TimothyWillard committed:
Add `subset_dataframe` to filter by compartments

Added the `Compartments.subset_dataframe` method to easily filter a
provided pandas DataFrame with specific column structure in a loop.

TimothyWillard committed:
Add `_invert_into_dict` internal helper

Added an internal helper, `_invert_into_dict`, to `gempyor.utils` to
invert a single value into multiple keys for a dictionary whose values
are a list of some type. This helper does this efficiently by modifying
the dictionary in place.

TimothyWillard committed:
Add 'Plugin' method to initial conditions

TimothyWillard committed:
Add `_inspect_requested_parameters` helper

Added `gempyor.parameters._inspect_requested_parameters` helper to
inspect requested parameters from a provided function. This function
will then extract those requested parameters from the given `pdata`
(structured like `Parameters.pdata` dict attribute) and `p_draw` numpy
array (like the return of `Parameters.parameters_quick_draw`). Also
added corresponding documentation and unit tests.

TimothyWillard committed:
Initial conditions plugins with parsed parameters

Updated the `create_initial_conditions` method, implemented by
subclasses of `InitialConditionsABC`, to accept `*args` for parsed
parameters instead of `parameters` and `p_draw` directly. These
requested parameters get parsed out in the `get_initial_conditions`
method used by `gempyor` to access initial conditions. Minor code
adjustments were made to `DefaultInitialConditions` and
`FileOrFolderDrawInitialConditions` to accommodate and a small test was
added for a custom plugin that requests a parameter argument.

TimothyWillard committed:
Initial conditions documentation

Updated the initial conditions documentation by:
* Split initial conditions vs seeding section into its own page,
* Heavily refactored existing initial condiditions guide,
* Added a table overview of the builtin methods,
* Documented how to implement a plugin and how to request parameters.

pearsonca commented on documentation/gitbook/gempyor/model-implementation/initial-conditions-vs-seeding.md:
```diff
@@ -0,0 +1,19 @@
+
+# Initial Conditions vs Seeding
+
+Both initial conditions and seeding can be used to create similar behaviors. While the previous pages cover in detail each component's functionality, below is a brief table highlighting the major differences to help you decide what to use for your situation.
```
This needs a bit of a tune-up. More like:

```suggestion
Infectious disease models require some infections to kick off transmission dynamics. These can be present in the starting state for the simulation (as `initial_conditions`) or introduced over time (via `seeding`). Other sections (links?) provide a detailed breakdown, but this table highlights the major differences between these approaches to help you decide what to use for your situation.
```

also, are initial conditions and seeding mutually exclusive?

also-also-aside, do we need to think about the parametrization-based approach for seeding as well?

pearsonca replied:
(i see from text below, yes, seeding still needs to enable parametrization)

TimothyWillard replied:
> also-also-aside, do we need to think about the parametrization-based approach for seeding as well?

I think at some point we might be interested in unifying the two concepts into one? Might be outside the scope of this PR though.

TimothyWillard replied:
> also, are initial conditions and seeding mutually exclusive?

No, these can be used in conjunction. They are both provided to `gempyor.steps_rk4.rk4_integration` and that function uses both.

pearsonca commented on documentation/gitbook/gempyor/model-implementation/initial-conditions-vs-seeding.md:
```diff
@@ -0,0 +1,19 @@
+
+|                         | Initial Conditions                                                                      | Seeding                                                                            |
+|-------------------------|-----------------------------------------------------------------------------------------|------------------------------------------------------------------------------------|
+| Main purpose            | Specifying compartments sizes at the initial time.                                      | Instantaneously changing compartment sizes at arbitrary times.                     |
```
to double check, but does seeding really instantly add to the infectious (or whatever) compartment? or is it generally taking an S and making an I? I think most often it should be modeling a force of exposure, which would mean that (for example) a more R population would experience fewer seeding introductions.

```suggestion
| Main purpose            | Specifying compartments sizes at the initial time.                                      | Instantaneously transferring between compartments at arbitrary times.                     |
```

TimothyWillard replied:
You're correct, I did not realize the difference between the two was important and got loose with the language. This is the code that takes care of the instantaneous transition:

https://github.com/HopkinsIDD/flepiMoP/blob/ad62f2d00082de5b60e62100be7b24fddc6931da/flepimop/gempyor_pkg/src/gempyor/steps_rk4.py#L230-L249

pearsonca commented on documentation/gitbook/gempyor/model-implementation/initial-conditions-vs-seeding.md:
```diff
@@ -0,0 +1,19 @@
+| Main purpose            | Specifying compartments sizes at the initial time.                                      | Instantaneously changing compartment sizes at arbitrary times.                     |
+| Default functionality   | Each subpopulations entire population is placed in the first compartment.               | No seeding events occur.                                                           |
+| Config section needed   | Optional, results in warning that default functionality used.                           | Optional.                                                                          |
+| Required input files    | Depending on method could be a parquet or CSV file(s).                                  | Yes, a CSV describing seeding events with a format dependent on method used.       |
```
Bit confusing - shouldn't we also be mentioning plugins here? Roughly, we support 1) a precalculated file (csv or parquet) that can be read into the necessary format OR 2) a script to calculate that necessary format, using model structure + parameters.

aside: i could also imagine supporting a simple funcational interface, as with other parameters.

TimothyWillard replied:
Does this cover that:

https://github.com/HopkinsIDD/flepiMoP/blob/ac61e062582c7d44714df8b264d8c0ffc1d6521a/documentation/gitbook/gempyor/model-implementation/initial-conditions-vs-seeding.md?plain=1#L18

TimothyWillard commented:
This PR has some size to it so I can see how it might be helpful to break it up. Clear breaks into separate PRs from what I can see are:

1. Misc utilities: https://github.com/HopkinsIDD/flepiMoP/pull/572/files/74b255dd60d3cbb720ebd44294602e61abe1fabe
2. `ModelMeta`, `TimeSetup`: https://github.com/HopkinsIDD/flepiMoP/pull/572/files/c1c88b361fdf7f0afd445b5ee5fa78d187d0530c..6a5aa03b4323aad44a20ba2b58c320197f1d7a41

pearsonca commented on documentation/gitbook/gempyor/model-implementation/specifying-initial-conditions.md:
````diff
@@ -1,192 +1,307 @@
-### `initial_conditions::method`
+```yaml
+initial_conditions:
+  method: Default
````
aside: weird that this is capitalized?

TimothyWillard replied:
This is maintaining the current behavior, can change if you would like.

pearsonca commented on documentation/gitbook/gempyor/model-implementation/specifying-initial-conditions.md:
```diff
@@ -1,192 +1,307 @@
+Would place the entire subpopulation into the 'S', 'child', 'unvaxxed' compartment. While 'S' usually makes sense for a model, declaring the entire subpopulation to be children is likely not reflective of reality. 
 
-#### SetInitialConditions
+### 'SetInitialConditions' Method
```
new minor issue: supporting both this and FromFile seems wasteful (and confusing?)

If we want to support two *schema* for the files, and then load based on what the column headers are, that seems fine. But different config methods seems likely to confuse user.

TimothyWillard replied:
Yeah, I see that, but this is just maintaining the current behavior. Would we want to change that?

pearsonca commented on documentation/gitbook/gempyor/model-implementation/specifying-initial-conditions.md:
```diff
@@ -1,192 +1,307 @@
+* `subpop_1`, `subpop_2`, etc. – One column for each different subpopulation, containing the value of the number of individuals in the described compartment in that subpopulation at the given date. Note that these are named after the node names defined by the user in the `geodata` file.
+* `date` – The calendar date in the simulation, in YYYY-MM-DD format. Only values with a date that matches to the simulation `start_date` will be used.
+
+### 'SetInitialConditionsFolderDraw' and 'InitialConditionsFolderDraw' Methods
```
related to collapsing the other file types, feels like this should also be introspected. if the user offers a folder instead of a specific file, do folder draw mode.

pearsonca commented on documentation/gitbook/gempyor/model-implementation/specifying-initial-conditions.md:
````diff
@@ -1,192 +1,307 @@
+
+In addition to the builtin methods users can also code their own initial conditions plugins. Under the hood the builtin initial condition methods are actually themselves plugins. To implement a plugin place a python file in the working directory, say `model_input/my_initial_conditions.py` for example, which contains a class that subclasses `gempyor.initial_conditions.InitialConditionsABC`. For ex ...
+
+```python
+from typing import Literal
+
+from gempyor.compartments import Compartments
+from gempyor.initial_conditions import (
+    InitialConditionsABC,
+    register_initial_conditions_plugin,
+)
+from gempyor.parameters import Parameters
+from gempyor.subpopulation_structure import SubpopulationStructure
+import numpy as np
+import numpy.typing as npt
+from pydantic import Field
+
+class TwoCompartmentInitialConditions(InitialConditionsABC):
+    method: Literal["TwoCompartment"] = "TwoCompartment"
+    weight: float = Field(gt=0.0, lt=1.0)
+    
+    def create_initial_conditions(
+        self,
+        sim_id: int,
+        compartments: Compartments,
+        subpopulation_structure: SubpopulationStructure,
+    ) -> npt.NDArray[np.float64]:
+        y0 = np.zeros((len(compartments.compartments), subpopulation_structure.nsubpops))
+        y0[0, :] = self.weight * subpopulation_structure.subpop_pop
+        y0[1, :] = (1.0 - self.weight) * subpopulation_structure.subpop_pop
+        return y0
+
+register_initial_conditions_plugin(TwoCompartmentInitialConditions)
+```
````
this might be a tall-ask, but: is there anyway to make this more like "define a method, with this name + this signature", and then have our machinery parse that on the fly into the proper plugin structure?

as written, this is asking a lot for people. anyway to use importlib here? ala, https://docs.python.org/3/library/importlib.html#importing-programmatically

pearsonca replied:
though, i do like the ability to specify multiple methods once and then switch between them depending upon config key.

TimothyWillard replied:
I could look into that, although this seems more appropriate for when we get to the stage of a general purpose plugin section? Might be a bit beyond the scope of this PR.

pearsonca commented on documentation/gitbook/gempyor/model-implementation/specifying-initial-conditions.md:
````diff
@@ -1,192 +1,307 @@
+register_initial_conditions_plugin(TwoCompartmentInitialConditions)
+```
+
+Note that in the above plugin an additional argument has been added to `create_initial_conditions` whose name, `alpha`, corresponds to the SEIR parameter name. You can finally use this plugin in your configuration file:
````
nice. also feels like we should be telling people here what happens if they typo `alfa` instead - presumably some "could not find parameter ..." error?

TimothyWillard replied:
That's here:

https://github.com/HopkinsIDD/flepiMoP/blob/ac61e062582c7d44714df8b264d8c0ffc1d6521a/flepimop/gempyor_pkg/src/gempyor/parameters.py#L416-L421

I can add it to the todo list to add this to the documentation with an example.

pearsonca commented on documentation/gitbook/gempyor/model-implementation/specifying-initial-conditions.md:
```diff
@@ -1,192 +1,307 @@
+initial_conditions:
+  method: TwoCompartment
+  module: model_input/my_initial_conditions.py
+  weight: 0.5
```
`weight` is a weird special parameter.

i need to think a bit about this one - is it holdover from old interface?

rather than a class field, should `weight` be handled more like ...you can put any variables that only your initial conditions script uses in this config section and they'll get passed in just like normal parameters?

TimothyWillard replied:
I think if we want to pass the configuration as arguments rather than having them as attributes of the class that would require a rethink of the current pydantic usage. I'll add it to the todo list to investigate.

Downside to switching to arguments would be you loose any config parse time validations, instead errors like `weight` being `-0.1` would only be discovered at runtime.

pearsonca commented on examples/simple_usa_statelevel/simple_usa_statelevel.yml:
```diff
@@ -11,8 +11,7 @@ subpop_setup:
 initial_conditions:
-  method: plugin
-  plugin_file_path: model_input/my_initial_conditions.py
+  method: Default
```
meant to change this example? or should there be another example that shows default vs plugin vs ...?

TimothyWillard replied:
Yes, meant to change temporarily since the way plugins are used is different.

TimothyWillard committed:
Fix multiprocessing issues with custom plugins

When using initial conditions custom plugins, like ones that can be
inferred, the module has to be related when the multiprocessing start
method is 'spawn' or 'forkserver' because the child process does not get
a copy of the parent process' memory and therefore does not have the
dynamically loaded plugin available at start. This can lead to an unsual
'ModuleNotFound' exception with a traceback relating to deserializing
the pickled memory state. Luckily, the fix is simple, just need to
reload the custom plugin when the child process starts. This should be
kept in mind, and likely refactored when addressing #421.

pearsonca reviewed (commented):
some work in progress notes here.

TimothyWillard committed:
Add/modify examples to demo IC fitting

Modified the example contained in `examples/simple_usa_statelevel/` to
infer the initial conditions with a custom plugin. This custom initial
conditions plugin is very simple, just figuring out what proportion of
the population to place into the susceptible vs the infected
compartments. However, this is a great way to introduce users into
inferable initial conditions.

Also added configuration file
`examples/tutorials/config_sample_2pop_inference_with_initial_conditions.yml`
to demonstrate how to do a similar procedure with R inference.

pearsonca commented on documentation/gitbook/gempyor/model-implementation/specifying-initial-conditions.md:
```diff
@@ -1,192 +1,341 @@
- initial_conditions:
-   method: default
+initial_conditions:
+  method: default
```
i'm fine w/ caps vs not (prefer not, but for now let's prioritize minimizing changes)

```suggestion
  method: Default
```

pearsonca commented on documentation/gitbook/gempyor/model-implementation/specifying-initial-conditions.md:
```diff
@@ -1,192 +1,341 @@
+  age_group: ["child", "adult"]
+ 
+initial_conditions:
+  method: default
```
```suggestion
  method: Default
```

TimothyWillard committed:
Modify R inference to handle IC plugins

Made minor adjustments to the `inference` R package to support initial
condition plugins. Mostly adjusting logic around which code path to go
down based on the `method` field of the `initial_conditions` section.

Also fixed a bug in the `flepimop-inference-main` command where it was
trying to call the `flepimop-inference-slot` command with `R` rather
than directly invoking it.

TimothyWillard commented:
This PR is blocked by #589.

pearsonca commented on flepimop/gempyor_pkg/src/gempyor/calibrate.py:
```diff
@@ -142,10 +138,10 @@ def calibrate(
     if input_run_id is None:
         base_run_id = pathlib.Path(config_filepath).stem.replace("config_", "")
-        run_id = f"{base_run_id}-{file_paths.run_id()}"
+        run_id = f"{base_run_id}-{file_paths_run_id()}"
```
this seems like a spurious change? the `.run` to `_run`?

TimothyWillard replied:
This one was not spurious: https://github.com/HopkinsIDD/flepiMoP/blob/ee05ff3cea6cc7dabdb0930c877529844758eb78/flepimop/gempyor_pkg/src/gempyor/calibrate.py#L21

I decided to change this to only import the required function rather than the whole `file_paths` module.

emprzy commented on documentation/gitbook/gempyor/model-implementation/specifying-initial-conditions.md:
```diff
@@ -1,192 +1,350 @@
 ## Overview
 
-In order for the models specified previously to be dynamically simulated, the user must provide _initial conditions,_ in addition to the model structure and parameter values. Initial conditions describe the value of each variable in the model at the time point that the simulation is to start. For example, on day zero of an outbreak, we may assume that the entire population is susceptible except f ...
+In order for the models specified previously to be dynamically simulated the state of the system at the initial time must be specified, this is done via initial conditions. Initial conditions describe the value of each variable in the model at the time point that the simulation is to start. For example, on day zero of an outbreak, we may assume that the entire population is susceptible except for ...
```
```suggestion
In order for the previously specified models to be dynamically simulated the state of the system at the initial time must be specified, which is done via initial conditions. Initial conditions describe the value of each variable in the model at the time point that the simulation is to start. For example, on day zero of an outbreak, we may assume that the entire population is susceptible except for one infected individual. Alternatively, we could assume that some portion of the population already has prior immunity due to vaccination or previous infection. Different initial conditions lead to different model trajectories.
```

Sorry that's really picky haha but for clarity and flow!

emprzy commented on documentation/gitbook/gempyor/model-implementation/specifying-initial-conditions.md:
```diff
@@ -1,192 +1,350 @@
+### 'SetInitialConditions' Method
 
-#### Default
+With this method users can specify arbitrary initial conditions in a convenient formatted input CSV or parquet file. For example, for a model with the following `compartments` and `initial_conditions` sections:
```
```suggestion
With this method users can specify arbitrary initial conditions in a conveniently formatted input CSV or parquet file. For example, for a model with the following `compartments` and `initial_conditions` sections:
```

emprzy commented on examples/tutorials/model_input/two_compartment_initial_conditions.py:
```diff
@@ -0,0 +1,33 @@
+        subpopulation_population = np.array(
+            subpopulation_structure.subpop_pop, dtype=np.float64
+        )
+        y0 = np.zeros((len(compartments.compartments), subpopulation_structure.nsubpops))
```
Perhaps this is happening somewhere else the code, but do we want to have a check/error thrown for when there are less than two compartments? Otherwise i think this will fail

TimothyWillard replied:
I think only having one compartment is an usual situation that will break other portions of the code base, but this should be fine:
```python
>>> import numpy as np
>>> np.zeros((1, 1)).shape
(1, 1)
```
However, it should definitely be a warning if not an error in the `Compartments` class if there's only one compartment if it's not already. A gempyor model with only one compartment just doesn't make sense from a modeling/conceptual standpoint.

emprzy replied:
Ok cool. That makes a lot of sense, thanks.

emprzy commented on flepimop/gempyor_pkg/src/gempyor/_multiprocessing.py:
```diff
@@ -0,0 +1,87 @@
+    """
+    if multiprocessing.get_start_method() in {"forkserver", "spawn"}:
+        for module_path in custom_initial_conditions:
+            module_name = module_path.stem
```
Is just the stem of the file unique enough?

TimothyWillard replied:
For now it is since only one initials conditions plugin can be loaded at a time at the moment, but this is a very good point to keep in mind when working on #421 since users will then be able to use multiple plugins at once.

emprzy commented on flepimop/gempyor_pkg/src/gempyor/calibrate.py:
```diff
@@ -161,7 +157,8 @@ def calibrate(
             return
         print(
-            f"Doing a resume from {filename}, this only work with the same number of slot/walkers and parameters right now"
+            f"Doing a resume from {filename}, this only work with the "
```
```suggestion
            f"Doing a resume from {filename}, this only works with the "
```

emprzy commented on flepimop/gempyor_pkg/src/gempyor/initial_conditions/_plugins.py:
```diff
@@ -0,0 +1,193 @@
+            f"'method' name of '{method}'. Instead the available "
+            f"options are: {', '.join(sorted(_initial_conditions_plugins.keys()))}."
+        )
+    return initial_conditions_class.from_confuse_config(
```
I know `.from_confuse_config()` was not created in this PR, but could be worth renaming to `.subpopstruct_from_confuse_config()` (or something like that) now that we have `.distribution_from_confuse_config()`. Thoughts?

TimothyWillard replied:
Now that I'm thinking it through, I think we want to lean in the opposite direction (i.e. make `distribution_from_confuse_config` a class method of `DistributionABC` and rename it to `from_confuse_config`). Especially if we keep leaning into this plugin goal for model components, they'll need consistent interfaces. No changes required for `gempyor.distributions` for now though, probably a #421 issue to consolidate this into a `GempyorComponent` ABC or similar.

pearsonca replied:
agree - a static / class method (basically, an input type specific constructor) makes a lot of sense.

then, once the great confuse purge is complete, we can 🔥 them all.

emprzy replied:
I didn't even realize we wanted to torch confuse...

pearsonca replied:
oh, definitely we do. confuse is aptly named to say the least.

emprzy commented on documentation/gitbook/SUMMARY.md:
```diff
@@ -11,6 +11,7 @@
   * [Specifying seeding](gempyor/model-implementation/specifying-seeding.md)
+  * [Initial Conditions vs Seeding](./gempyor/model-implementation/initial-conditions-vs-seeding.md)
   * [Specifying observational model](gempyor/model-implementation/outcomes-for-compartments.md)
   * [Distributions](gempyor/model-implementation/distributions.md)
```
I'm going to make an issue to update the distributions documentation!

TimothyWillard replied:
#596

pearsonca committed:
expand input file explanation for initial conditions

pearsonca committed:
tweaking Default intial_conditions description
